internal/metrics/prometheus.go

```diff
@@ -0,0 +1,159 @@
+)
+
+// Recorder is the interface fir recording metrics on a certain operation.
+type Recorder interface {
+	// REVIEW: add URL of Konnect API as a label?
+	RecordKonnectEntityOperationSuccess(operationType KonnectEntityOperation, entityType string, duration time.Duration)
+	RecordKonnectEntityOperationFailure(operationType KonnectEntityOperation, entityType string, duration time.Duration, statusCode int)
+}
+
+type KonnectEntityOperation string
+
+const (
+	KonnectEntityOperationTypeKey                        = "operation_type"
+	KonnectEntityOperationCreate  KonnectEntityOperation = "create"
+	KonnectEnttiyOperationUpdate  KonnectEntityOperation = "update"
+	KonnectEntityOperationDelete  KonnectEntityOperation = "delete"
+
+	KonnectEntityTypeKey = "entity_type"
+
+	SuccessKey   = "success"
+	SuccessTrue  = "true"
+	SuccessFalse = "false"
+
+	StatusCodeKey = "status_code"
+)
+
+// metric names for konnect entity operations.
+const (
+	// REVIEW: define a Namespace `gateway_operator` for creating prometheus metrics here?
+	MetricNameKonnectEntityOperationCount    = "gateway_operator_konnect_entity_operation_count"
+	MetricNameKonnectEntityOperationDuration = "gateway_operator_konnect_entity_operation_duration"
+)
+
+var (
+	konnectEntityOperationCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: MetricNameKonnectEntityOperationCount,
+			Help: fmt.Sprintf(
+				"Count of successful/failed entity operations in Konnect. "+
+					"`%s` describes the operation type (`%s`, `%s`, or `%s`)."+
+					"`%s` describes the type of the operated entity. "+
+					"`%s` describes whether the operation is successful (`%s`) or not (`%s`). "+
+					"`%s` is populated in case of `%s=\"%s\"` and describes the status code returned from Konnect API.",
+				KonnectEntityOperationTypeKey, KonnectEntityOperationCreate, KonnectEnttiyOperationUpdate, KonnectEntityOperationDelete,
+				KonnectEntityTypeKey,
+				SuccessKey, SuccessTrue, SuccessFalse,
+				StatusCodeKey, SuccessKey, SuccessFalse,
+			),
+		},
+		[]string{KonnectEntityOperationTypeKey, KonnectEntityTypeKey, SuccessKey, StatusCodeKey},
+	)
+
+	konnectEntityOperationDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name: MetricNameKonnectEntityOperationDuration,
+			Help: fmt.Sprintf(
+				"How long did the Konnect entity operation take in seconds. "+
+					"`%s` describes the operation type (`%s`, `%s`, or `%s`)."+
+					"`%s` describes the type of the operated entity. "+
+					"`%s` describes whether the operation is successful (`%s`) or not (`%s`). "+
+					"`%s` is populated in case of `%s=\"%s\"` and describes the status code returned from Konnect API.",
+				KonnectEntityOperationTypeKey, KonnectEntityOperationCreate, KonnectEnttiyOperationUpdate, KonnectEntityOperationDelete,
+				KonnectEntityTypeKey,
+				SuccessKey, SuccessTrue, SuccessFalse,
+				StatusCodeKey, SuccessKey, SuccessFalse,
+			),
+			// Duration range from 0.1s to 10min (600s).
```

Comment:
0.1s being 100ms is to big for minimal value. I think 1ms would be good to use in this case. WDYT?

Reply:
The operations involve API calls to an external cloud service so I think 1ms is too small (It takes ~200ms in my tests). So I think considering the cases with shorter network latency, 10ms should be a proper minimum value.

Reply:
Changed to 1ms~10min.